File: exercicios/projeto-final/skeletons/rickandmorty/flutter/pratica/lib/screens/detail_screen.dart
import 'package:flutter/material.dart';
import '../api/rickandmorty_api.dart';
import '../models/character.dart';
import '../storage/favorites_store.dart';

class DetailScreen extends StatefulWidget {
  final int characterId;

  const DetailScreen({super.key, required this.characterId});

  @override
  State<DetailScreen> createState() => _DetailScreenState();
}

class _DetailScreenState extends State<DetailScreen> {
  final RickAndMortyApi _api = RickAndMortyApi();
  final FavoritesStore _favoritesStore = FavoritesStore();

  CharacterDetail? _detail;
  bool _isFavorite = false;
  String? _error;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    // TODO 2 (feature 2 — detalhe): chamar `_api.fetchDetail(widget.characterId)`
    // e `_favoritesStore.load()`, guardar em `_detail`/`_isFavorite` via
    // setState. Tratar erro com `_error` (ver padrão do catch abaixo).
  }

  Future<void> _toggleFavorite() async {
    // TODO 5 (feature 5 — favoritos): chamar `_favoritesStore.toggle(widget.characterId)`
    // e atualizar `_isFavorite` via setState.
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: Semantics(
          identifier: 'detail-back-button',
          child: IconButton(
            icon: const Icon(Icons.arrow_back),
            onPressed: () => Navigator.of(context).pop(),
          ),
        ),
        title: const Text('Detalhe'),
      ),
      body: Semantics(
        identifier: 'detail-screen',
        child: _buildBody(),
      ),
    );
  }

  Widget _buildBody() {
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    final detail = _detail;
    if (detail == null) {
      return const Center(child: CircularProgressIndicator());
    }
    return Padding(
      padding: const EdgeInsets.all(16),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Center(
            child: Image.network(detail.imageUrl, height: 140),
          ),
          const SizedBox(height: 16),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Semantics(
                identifier: 'detail-title',
                child: Text(
                  capitalize(detail.name),
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ),
              Semantics(
                identifier: 'detail-favorite-button',
                child: IconButton(
                  icon: Icon(
                    _isFavorite ? Icons.favorite : Icons.favorite_border,
                    color: Colors.redAccent,
                  ),
                  onPressed: _toggleFavorite,
                ),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Chip(label: Text(detail.status)),
          const SizedBox(height: 16),
          Text('Espécie: ${detail.species}'),
          Text('Gênero: ${detail.gender}'),
        ],
      ),
    );
  }
}

File: exercicios/projeto-final/skeletons/rickandmorty/kmp/pratica/androidApp/src/main/kotlin/com/puciec/rickmortykmp/DetailScreen.kt
package com.puciec.rickmortykmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.rickmortykmp.data.RickAndMortyApi
import com.puciec.rickmortykmp.data.CharacterDetail
import com.puciec.rickmortykmp.data.capitalize

@Composable
fun DetailScreen(
    api: RickAndMortyApi,
    favoritesStore: FavoritesStore,
    characterId: Int,
    onBack: () -> Unit,
) {
    var detail by remember { mutableStateOf<CharacterDetail?>(null) }
    var isFavorite by remember { mutableStateOf(favoritesStore.load().contains(characterId)) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(characterId) {
        try {
            detail = api.fetchDetail(characterId)
        } catch (t: Throwable) {
            error = "Não foi possível carregar o detalhe."
        }
    }

    Column(Modifier.fillMaxSize().testTag("detail-screen").padding(16.dp)) {
        Text(
            text = "< Voltar",
            modifier = Modifier.testTag("detail-back-button").clickable { onBack() },
        )
        when {
            error != null -> Text(error!!)
            detail == null -> Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }
            else -> {
                val d = detail!!
                Row(
                    Modifier.fillMaxWidth().padding(top = 16.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Text(
                        text = capitalize(d.name),
                        style = MaterialTheme.typography.headlineMedium,
                        modifier = Modifier.testTag("detail-title"),
                    )
                    Text(
                        text = if (isFavorite) "♥" else "♡",
                        modifier = Modifier
                            .testTag("detail-favorite-button")
                            .clickable {
                                val ids = favoritesStore.toggle(characterId)
                                isFavorite = ids.contains(characterId)
                            },
                        style = MaterialTheme.typography.headlineMedium,
                    )
                }
                Text(d.status)
                Text("Espécie: ${d.species}")
                Text("Gênero: ${d.gender}")
            }
        }
    }
}
